### PR TITLE
chore: bump Lean to v4.9.0

### PR DESCRIPTION
**Pull request by joewatt95**
This bumps Lean and Mathlib to the latest (as of 2024-7-6) stable version.

### CODE REVIEW

File: Smt/Reconstruct/Real/Polynorm.lean [hunk 1]
```diff
@@ -5,7 +5,8 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Abdalrhman Mohamed
 -/
 
-import Mathlib.Data.Real.Basic
+import Mathlib.Data.Real.Star
```
Comment (joewatt95) [inline Smt/Reconstruct/Real/Polynorm.lean:8]: This change is needed as `Real.Basic` as per `4.8.0` has been split up to `Real.Basic` and `Real.Star` as per `4.9.0`. Reverting this results in the following error when tests are run:
```
Test/linarith.lean:146:2: error: [arithPolyNorm]: could not prove x - y = -x + y
```

File: Smt/Data/Sexp.lean [hunk 2]
```diff
@@ -83,9 +83,9 @@ partial def parseOneAux : List Substring → Except ParseError (Sexp × List Sub
       return (atom tk.toString, tks)
   | [] => throw <| .incomplete "expected a token, got none"
 
-partial def parseManyAux :=
+partial def parseManyAux : List Substring → Except ParseError (Array Sexp × List Substring) :=
```
Comment (joewatt95) [inline Smt/Data/Sexp.lean:86]: Shifting the type annotation up from `go` to here helps resolves error messages involved with type inference / checking, which we get with `4.9.0`.

File: Smt/Reconstruct/Real/Polynorm.lean (hunk 1 above)
Comment (abdoo8080) [reply Smt/Reconstruct/Real/Polynorm.lean:8]: Is the error above the only one caused by the version change? If so, then let's keep `import Mathlib.Data.Real.Basic`. This error message is an instance of a bigger issue that will be resolved with once [cvc5/#10911](https://github.com/cvc5/cvc5/pull/10911) is merged.

**Review by abdoo8080 — COMMENTED** [review]
Hi @joewatt95! Thanks for the PR! It looks good to me. I left just one minor comment.

File: Smt/Reconstruct/Real/Polynorm.lean [hunk 3]
```diff
@@ -5,7 +5,8 @@ Released under Apache 2.0 license as described in the file LICENSE.
...
+import Mathlib.Tactic.Ring.RingNF
```
Comment (joewatt95) [inline Smt/Reconstruct/Real/Polynorm.lean:9]: Without this, compilation fails with the following error:
```
unknown namespace 'Mathlib.Tactic.RingNF'
```
It looks like `Mathlib.Data.Real.Basic` (which used to include `Mathlib.Data.Real.Star`) transitively imported it, but no longer does so now.

File: Smt/Reconstruct/Real/Polynorm.lean (hunk 1 above)
Comment (joewatt95) [reply Smt/Reconstruct/Real/Polynorm.lean:8]: Yep, this is the only failing test case caused by the version change. I'll revert the change to this import following your suggestion here, thank you.

Also, thank you for mentioning the related cvc5 issue. I found this test case breakage to be rather puzzling, and suspected that there may be a deeper issue involving cvc5's internals.

**Review by abdoo8080 — APPROVED** [review]

**Pull request merged**